Initial v2 record creation

chdwlch opened the pull request:
Initial implementation of https://github.com/ATLBitLab/twelvecash/issues/10
Adds `/v2/record` route that takes additional parameters for the construction of the bip21 uri.
Also sets up some environment variable checks and basic unit tests with jest

New record route allows for custom query parameters to support things like lnurl.
It also accepts a domain if we want to support multiple domains.

```
# dig txt chad.user._bitcoin-payment.regtest.12cash.dev
...
;; ANSWER SECTION:
chad.user._bitcoin-payment.regtest.12cash.dev. 3600 IN TXT "bitcoin:bc1...?lno=lno1...&sp=sp1..."
```

chdwlch commented:
Switched the NETWORK env var to use "mainnet" instead of being undefined to make the environment validator more straight forward. Will need that to be changed in the netlify settings

sbddesign reviewed (commented):
Looking really awesome. I have some questions and nits. Main thing I am concerned about is the formatting of the error response.

sbddesign commented on src/app/v2/record/route.ts:
Also, I tried stuffing excess data and got a response of `{"error":{}}`. Seems like the error could be more informative.

```
{
    "domain": "12cash.dev",
    "username": "test202407141305edtb",
    "lno": "lno123...xyz",
    "sp": "sp123...xyz",
    "onChain": "bc1p...xyz",
    "custom": [
        {
            "prefix": "lnurl",
            "value": "lnur123...xyz"
        },
        {
            "prefix": "veggie",
            "value": "carrot"
        },
        {
            "prefix": "mystory",
            "value": "Bitcoin ipsum dolor sit amet. Block reward double-spend problem peer-to-peer nonce blockchain, difficulty stacking sats Bitcoin Improvement Proposal? Hard fork consensus UTXO SHA-256 mining wallet outputs whitepaper? Proof-of-work hash proof-of-work hashrate blocksize outputs halvening inputs? Proof-of-work stacking sats Bitcoin Improvement Proposal whitepaper nonce, hash hashrate. Blocksize double-spend problem hashrate Bitcoin Improvement Proposal sats segwit SHA-256 whitepaper transaction! Miner double-spend problem digital signature block height blockchain soft fork stacking sats? Hashrate key pair, mining address stacking sats hashrate mempool, address. Proof-of-work UTXO peer-to-peer hash difficulty. Mining sats blockchain key pair sats UTXO mining. Mempool full node cryptocurrency SHA-256 wallet stacking sats consensus segwit. UTXO proof-of-work sats transaction difficulty genesis block stacking sats address. Double-spend problem whitepaper hash, consensus blocksize mempool Merkle Tree! Bitcoin Improvement Proposal outputs consensus full node segwit peer-to-peer bitcoin. Timestamp server, transaction miner, full node peer-to-peer blocksize digital signature. Stacking sats double-spend problem Merkle Tree mempool blocksize decentralized key pair mining Satoshi Nakamoto? Transaction decentralized Bitcoin Improvement Proposal, satoshis transaction double-spend problem stacking sats peer-to-peer. Block height. Hash block height public key whitepaper, key pair hashrate, soft fork. Blocksize satoshis outputs sats SHA-256 genesis block mempool miner? Timestamp server mempool key pair double-spend problem, Merkle Tree whitepaper decentralized hash, whitepaper. Bitcoin Improvement Proposal, blockchain block reward mining hashrate block reward cryptocurrency block reward. Whitepaper, private key sats. Proof-of-work soft fork hard fork mining peer-to-peer stacking sats halvening wallet. Difficulty hard fork blockchain mempool block height address, satoshis inputs mempool! Cryptocurrency stacking sats blocksize soft fork private key soft fork decentralized double-spend problem! Nonce, blocksize! Merkle Tree Satoshi Nakamoto stacking sats full node public key bitcoin, mining, miner. Genesis block whitepaper bitcoin timestamp server outputs digital signature full node, decentralized. Transaction mining mempool, timestamp server peer-to-peer sats UTXO. Miner genesis block whitepaper Merkle Tree hard fork hash transaction electronic cash. Cryptocurrency hashrate proof-of-work blockchain outputs block reward outputs hash. Blocksize hard fork proof-of-work hard fork genesis block private key genesis block. Hashrate wallet address outputs hash, hard fork, halvening? Hard fork UTXO outputs, blocksize UTXO genesis block mining. Full node SHA-256, outputs, segwit Merkle Tree UTXO address, proof-of-work consensus. Consensus wallet, sats double-spend problem digital signature consensus blocksize transaction whitepaper. Halvening block reward blocksize, hard fork Satoshi Nakamoto timestamp server proof-of-work peer-to-peer. Miner nonce? Difficulty digital signature miner nonce proof-of-work consensus cryptocurrency halvening key pair? Cryptocurrency hash block reward halvening mining genesis block, hash blocksize bitcoin. Satoshis mempool, consensus private key, hard fork, whitepaper UTXO. Consensus full node Merkle Tree sats timestamp server consensus miner, segwit. Double-spend problem digital signature."
        }
    ]
}
```

sbddesign replied:
Works now

chdwlch commented on .env.local.sample:
```diff
@@ -11,5 +11,5 @@ PROVIDER="cloudflare"
-# Network: leave empty for mainnet, "testnet" for testnet, "regtest" for regtest
-#NETWORK="regtest"
\ No newline at end of file
+# Network: "mainnet" for mainnet, "testnet" for testnet, "regtest" for regtest
```
NETWORK just needed to be defined so the `env.js` validation would pass, so i set it to "mainnet"
I can set it to an empty string for mainnet

chdwlch commented on env.js:
```diff
@@ -0,0 +1,28 @@
+// @ts-check
+/**
+ * This file is included in `/next.config.js` which ensures the app isn't built with invalid env vars.
+ * It has to be a `.js`-file to be imported there.
+ */
+/* eslint-disable @typescript-eslint/no-var-requires */
```
I think it just runs when it's imported into `const { env } = require('./env')`
I copied most of these config files and stuff from my other project so there is a bunch of comments and things that might not be needed in this project.
I'll clean them up

chdwlch commented on src/app/v2/record/route.ts:
Yeah i need to fix the error outputs still

chdwlch commented on src/app/v2/record/route.ts:
Hmm it's working for me

```
curl -X POST http://localhost:3000/v2/record -H "Content-Type: application/json" -d  '{"domain": "12cash.dev","username": "test2024chad","lno": "lno123...xyz","sp": "sp123...xyz","onChain": "bc1p...xyz","custom": [{"prefix": "lnurl","value": "lnur123...xyz"},{"prefix": "veggie","value": "carrot"}]}'


test2024chad.user._bitcoin-payment.12cash.dev. 3600 IN TXT "bitcoin:bc1p...xyz?lno=lno123...xyz&sp=sp123...xyz&lnurl=lnur123...xyz&veggie=carrot"
```

sbddesign replied:
Works for me now

chdwlch commented on src/app/v2/record/route.ts:
```diff
@@ -0,0 +1,106 @@
+import { z } from "zod";
+const axios = require("axios").default;
+
+const DOMAINS = ["12cash.dev"] as const;
```
Yeah i need to refactor that still to support multiple domains in the env.
Trying to figure out the right way to do it since every additional domain requires a different CF domain ID

chdwlch commented:
Fixed the bad formatted error messages and switched them all to use a more consistent structure.

```
{"error":{"message":"Failed to create bip21 URI","errors":[{"message":"Bip21 URI is greater than 2048 characters"}]}}

{"error":{"message":"Invalid request","errors":[{"code":"custom","message":"At least one payment option is required","path":[]}]}}

{"error":{"message":"Invalid request","errors":[{"code":"invalid_type","expected":"string","received":"undefined","path":["userName"],"message":"Required"}]}}
```

switched username to userName

sbddesign reviewed (approved):
LGTM
